docs/advanced/authentication.md

```diff
@@ -0,0 +1,328 @@
+  }
+
+  return null;
+}
+```
+
+We check that the `authState` doesn't already exist (this indicates that it is the first time this exchange is executed and not an auth failure) and fetch the auth state from
+storage. The structure of this particular`authState` is an object with keys for `token` and `refreshToken`, but this format is not required. You can
+use different keys or store any additional auth related information here. For example you could decode and store the token expiry date, which would save you from decoding
+your JWT every time you want to check whether your token is expired.
+
+In React Native, this is very similar, but because persisted storage in React Native is always asynchronous, so is this function:
+
+```js
+const getAuth = async ({ authState, mutate }) => {
+  if (!authState) {
+    const token = await AsyncStorage.getItem(TOKEN_KEY, {});
+    const refreshToken = await AsyncStorage.getItem(REFRESH_TOKEN_KEY, {});
+    if (token && refreshToken) {
+      return { token, refreshToken };
+    }
+    return null;
+  }
+
+  return null;
+}
+```
+
+### Configuring `addAuthToOperation`
+
+The purpose of `addAuthToOperation` is to take apply your auth state to each request. Note that the format of the `authState` will be whatever
+you've returned from `getAuth` and not at all constrained by the exchange:
+
+```js
+const addAuthToOperation = ({
+  authState,
+  operation,
+}) => {
+  if (!authState || !authState.token) {
+    return operation;
+  }
+
+  const fetchOptions =
+    typeof operation.context.fetchOptions === 'function'
+      ? operation.context.fetchOptions()
+      : operation.context.fetchOptions || {};
+
+  return {
+    ...operation,
+    context: {
+      ...operation.context,
+      fetchOptions: {
+        ...fetchOptions,
+        headers: {
+          ...fetchOptions.headers,
+          "Authorization": authState.token,
+        },
+      },
+    },
+  };
+}
+```
+
+First we check that we have an `authState` and a `token`. Then we apply it to the request `fetchOptions` as an `Authorization` header.
+The header format can vary based on the API (e.g using `Bearer ${token}` instead of just `token`) which is why it'll be up to you to add the header
+in the expected format for your API.
+
+### Configuring `didAuthError`
+
+This function lets the exchange know what is defined to be an API error for your API:
+
+```js
+const didAuthError = ({ error }) => {
+  return error.graphQLErrors.some(
+    e => e.extensions?.code === 'FORBIDDEN',
+  );
+}
+```
+
+For most graphQL APIs, the auth error is communicated as an error code, however it may also be a 401 HTTP response:
+
+```js
+const didAuthError = ({ error }) => {
+  return error.graphQLErrors.some(
+    e => error.response.status === 401,
+  );
+},
+```
+
+Then `didAuthError` returns `true`, it will trigger the exchange to trigger the logic for asking for reauthentication via `getAuth`.
+
+### Configuring `getAuth` (after `didAuthError` returns `true`)
```

Comment:
If "after `didAuthError` returns `true` is important information and there's no corresponding section that for instance says `false` instead, let's move this into the text below as a full explanation?

> `getAuth` will be triggered to refresh the authentication state if `didAuthError` returned `true` previously, due to at least one GraphQL request erroring out with an authentication error, as we defined previously.

Reply:
The two sections are "Configuring `getAuth` (fetch from storage)" and "Configuring `getAuth` (after `didAuthError` returns `true`)".

I could rephrase it to "after an auth error has happened"?